pallets/deitos/src/lib.rs

```diff
@@ -113,54 +135,180 @@ pub mod pallet {
+
+            Self::deposit_event(Event::IPRegistered {
+                ip,
+                price_storage_per_block,
+                total_storage,
+            });
 
             Ok(())
         }
 
+        /// This is a temporary call to manage the IP status.
+        /// Statuses updates should be done automatically after an environment software check.
         #[pallet::call_index(1)]
-        #[pallet::weight(T::WeightInfo::update_ip_details())]
-        pub fn update_ip_details(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
-            // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+        #[pallet::weight(T::WeightInfo::register_ip())]
+        pub fn update_ip_status(
+            origin: OriginFor<T>,
+            provider: AccountIdLookupOf<T>,
+            status: IPStatus,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let bounding_amount = IPBoundingAmount::<T>::get()?;
+            let ip = T::Lookup::lookup(provider)?;
+
+            InfrastructureProvider::<T>::try_mutate(
+                &ip,
+                |ip_details| -> Result<_, DispatchError> {
+                    let ip_details = ip_details
+                        .as_mut()
+                        .ok_or(Error::<T>::NonExistentStorageValue)?;
+
+                    ip_details.status = status.clone();
+
+                    Self::deposit_event(Event::IPStatusChanged {
+                        ip: ip.clone(),
+                        status,
+                    });
+
+                    Ok(())
+                },
+            )?;
 
             Ok(())
         }
 
         #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::update_ip_storage())]
+        pub fn update_ip_storage(
+            origin: OriginFor<T>,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        ) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer.
+            let ip = ensure_signed(origin)?;
+
+            InfrastructureProvider::<T>::try_mutate(
+                &ip,
+                |ip_details| -> Result<_, DispatchError> {
+                    let ip_details = ip_details
+                        .as_mut()
+                        .ok_or(Error::<T>::NonExistentStorageValue)?;
+
+                    // Check if the new total_storage is enough to cover the current reserved_storage
+                    ensure!(
+                        total_storage >= ip_details.reserved_storage,
+                        Error::<T>::InsufficientStorage
+                    );
+
+                    ip_details.price_storage_per_block = price_storage_per_block;
+                    ip_details.total_storage = total_storage;
+                    // Note: This calculation might need to be reviewed or adjusted.
+                    ip_details.reserved_storage = ip_details
```

Comment:
Does `reserved_storage` mean the storage currently in use? Then I don't understand why does it suppose to change when IP updates the total storage.

Reply:
oh! good catch .. this is a huge mistake ... the reason why this is being updated is because i initially started to track `available_storage` and then i changed my mind to register `reserved_storage`. This sub should be deleted.